### PR TITLE
refactor(ava/advisor): migrate chart-advisor to v3 advisor

### PR DESCRIPTION
**Pull request by GuangMingYouBei**
1. TODO: unified type export

### CODE REVIEW

File: packages/ava/src/advisor/ruler/rules/aggregation-single-row.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
-import { CKBJson } from '@antv/ckb';
+import { ckb } from '../../../ckb';
 
 import type { RuleModule } from '../interface';
 
-const Wiki = CKBJson('en-US', true);
+const Wiki = ckb();
```
Comment (neoddish) [inline packages/ava/src/advisor/ruler/rules/aggregation-single-row.ts:5]: CKB should be param of the rule function, o/w. custom CKB could not work

Comment (neoddish) [reply packages/ava/src/advisor/ruler/rules/aggregation-single-row.ts:5]: same for all rule files

File: packages/ava/src/advisor/ruler/rules/aggregation-single-row.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
 const allChartTypes = Object.keys(Wiki) as string[];
 export const aggregationSingleRow: RuleModule = {
```
Comment (neoddish) [inline packages/ava/src/advisor/ruler/rules/aggregation-single-row.ts:7]: export function

File: packages/ava/src/advisor/ruler/interface.ts [hunk 3]
```diff
@@ -1,7 +1,12 @@
-import { analyzer } from '@antv/data-wizard';
+/**
```
Comment (neoddish) [inline packages/ava/src/advisor/ruler/interface.ts:1]:
```suggestion
/*
```
this is a js multi-line comment for file, not jsdoc

and,the name of this file should be `types.ts`

File: packages/ava/src/advisor/advisor.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  LintParams,
+  ScoringResultForChartType,
+  ScoringResultForRule,
+  Lint,
+} from './types';
+
+type AdviseLog = ScoringResultForChartType[];
+type LintLog = ScoringResultForRule[];
+
+/**
```
Comment (neoddish) [inline packages/ava/src/advisor/advisor.ts:25]:
```suggestion
/*
* TODO:
```

File: packages/ava/src/advisor/advisor.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * 0. 架构设计
+ * 1. Advisor 搬运
+ * 2. Linter 搬运
+ */
+
+export class Advisor {
+  /**
+   * CKB used for advising
+   */
+  CKB: ChartKnowledgeBase;
```
Comment (neoddish) [inline packages/ava/src/advisor/advisor.ts:36]:
```suggestion
  ckb: ChartKnowledgeBase;
```

File: packages/ava/src/advisor/advise-pipeline/data-to-advices.ts [hunk 6]
```diff
@@ -1,27 +1,32 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (neoddish) [inline packages/ava/src/advisor/advise-pipeline/data-to-advices.ts:1]:
```suggestion
// TODO
// eslint-disable-next-line import/no-extraneous-dependencies
```

File: packages/ava/src/advisor/advise-pipeline/data-to-advices.ts [hunk 7]
```diff
@@ -203,14 +208,14 @@ function applySmartColor(
  * @returns chart list [ { type: chartTypes, spec: antv-spec, score: >0 }, ... ]
  */
 export function dataToAdvices(
-  data: DataRows,
+  data: Datum[],
```
Comment (neoddish) [inline packages/ava/src/advisor/advise-pipeline/data-to-advices.ts:211]:
```suggestion
  data: Data,
```
`import type { Specification, Data } from '../../common/types';`

File: packages/ava/src/advisor/advise-pipeline/index.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+import { DataFrame } from '../../data';
+import { cloneDeep } from '../utils';
+
+import { dataToAdvices } from './data-to-advices';
+
+import type { AdviseResult, ChartAdviseParams } from '../types';
+import type { BasicDataPropertyForAdvice } from '../ruler';
+
+export function advicesForChart(params: ChartAdviseParams, exportLog = false): AdviseResult {
```
Comment (neoddish) [inline packages/ava/src/advisor/advise-pipeline/index.ts:14]: add a file advicesForChart.ts

and import the func to index

File: packages/ava/src/advisor/advise-pipeline/spec-mapping.ts [hunk 9]
```diff
@@ -37,7 +41,7 @@ function splitAngleColor(dataProps: BasicDataPropertyForAdvice[]): [ReturnField,
   return [field4Color, field4Angle];
 }
 
-function pieChart(data: DataRows, dataProps: BasicDataPropertyForAdvice[]): Advice['spec'] {
+function pieChart(data: Datum[], dataProps: BasicDataPropertyForAdvice[]): Advice['spec'] {
```
Comment (neoddish) [inline packages/ava/src/advisor/advise-pipeline/spec-mapping.ts:44]:
```suggestion
function pieChart(data:Data, dataProps: BasicDataPropertyForAdvice[]): Advice['spec'] {
```

File: packages/ava/src/advisor/advise-pipeline/spec-mapping.ts [hunk 10]
```diff
@@ -1,15 +1,19 @@
-import { CHART_ID_OPTIONS } from '@antv/ckb';
-import { statistics, utils as dwUtils } from '@antv/data-wizard';
+/**
+ * TODO
+ * -- statistics 改为从 dw 直接引入
+ */
```
Comment (neoddish) [inline packages/ava/src/advisor/advise-pipeline/spec-mapping.ts:4]:
```suggestion
/*
 * TODO
 * -- statistics 改为从 dw 直接引入
 * 在 CKB 中增加 encode 相关描述，让这里的 spec mapping 简化
 */
```

File: packages/ava/src/advisor/lint-pipeline/lintRules.ts [hunk 11]
```diff
@@ -1,17 +1,17 @@
...
+import type { Specification } from '../../common/types';
+import type { ScoringResultForRule, Lint } from '../types';
+import type { ChartRuleModule, DesignRuleModule } from '../ruler';
 
-const lintRules = (
+export function lintRules(
   ruleBase: Record<string, RuleModule>,
   ruleTypeToLint: 'notDESIGN' | 'DESIGN',
   info: Info,
   log: ScoringResultForRule[],
```
Comment (neoddish) [inline packages/ava/src/advisor/lint-pipeline/lintRules.ts:11]: mark todo: remove log param and console, add a func to return log

**Review by neoddish — CHANGES_REQUESTED** [review]

File: packages/ava/src/advisor/utils/index.ts [hunk 12]
```diff
@@ -1,4 +1,7 @@
 export { default as deepMix } from './deep-mix';
```
[the remark this answers is not in the document]
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/utils/index.ts:1]: done

File: packages/ava/src/advisor/types.ts [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+  fix?: any;
+  /** docs: explanation */
+  docs?: any;
+};
+
+export type AdviseParams = ChartAdviseParams;
+
+export type ChartAdviseParams = {
+  /** input data to advise */
+  data: Datum[];
```
[the remark this answers is not in the document]
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/types.ts:68]: done

[the remark this answers is not in the document]
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/types.ts:68]: done

File: packages/ava/src/advisor/utils/index.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/utils/index.ts:1]: done

File: packages/ava/src/advisor/ruler/rules/aggregation-single-row.ts (hunk 1 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/ruler/rules/aggregation-single-row.ts:5]: done

File: packages/ava/src/advisor/ruler/interface.ts (hunk 3 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/ruler/interface.ts:1]: done

File: packages/ava/src/advisor/advisor.ts (hunk 4 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/advisor.ts:25]: done

File: packages/ava/src/advisor/advisor.ts (hunk 5 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/advisor.ts:36]: done

File: packages/ava/src/advisor/advise-pipeline/data-to-advices.ts (hunk 6 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/advise-pipeline/data-to-advices.ts:1]: done

File: packages/ava/src/advisor/advise-pipeline/data-to-advices.ts (hunk 7 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/advise-pipeline/data-to-advices.ts:211]: done

File: packages/ava/src/advisor/advise-pipeline/index.ts (hunk 8 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/advise-pipeline/index.ts:14]: done

File: packages/ava/src/advisor/advise-pipeline/spec-mapping.ts (hunk 9 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/advise-pipeline/spec-mapping.ts:44]: done

File: packages/ava/src/advisor/advise-pipeline/spec-mapping.ts (hunk 10 above)
Comment (GuangMingYouBei) [reply packages/ava/src/advisor/advise-pipeline/spec-mapping.ts:4]: done

File: package.json [hunk 14]
```diff
@@ -85,9 +86,14 @@
   "dependencies": {
     "@antv/smart-color": "^0.2.1",
     "@rollup/plugin-commonjs": "^20.0.0",
+    "@rollup/plugin-json": "^4.1.0",
     "@rollup/plugin-node-resolve": "^13.0.4",
     "@rollup/plugin-typescript": "^8.2.5",
+    "lodash": "^4.17.21",
```
Comment (neoddish) [inline package.json:92]: rm smart-color & lodash

File: packages/ava/src/advisor/utils/index.ts [hunk 15]
```diff
@@ -1,4 +1,7 @@
...
 export { default as compare } from './compare';
 export { default as cloneDeep } from './clone';
+// eslint-disable-next-line import/no-cycle
```
[the remark this answers is not in the document]
Comment (neoddish) [reply packages/ava/src/advisor/utils/index.ts:8]: https://github.com/antvis/AVA/issues/502

**Review by neoddish — APPROVED** [review]
lgtm

**Pull request merged**